File: modules/flx_polar_h7/inspector/an-inspector/ForgeModule/src/io/trigger/forge/android/modules/flx_polar_h7/DataUploader.java
package io.trigger.forge.android.modules.flx_polar_h7;

import io.trigger.forge.android.core.ForgeApp;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.params.BasicHttpParams;
import org.apache.http.params.HttpConnectionParams;
import org.apache.http.params.HttpParams;
import org.apache.http.protocol.HTTP;
import org.json.JSONObject;

import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.util.Log;

/**
 * This class receives heart rate data to be uploaded, and uploads it within a separated thread.
 * 
 * @author Julien Dupuis
 */
public class DataUploader implements Runnable {
	
	// Number of data samples to accumulate before a flush
	public static final int BUNCH_SIZE = 30;
	
	// Max number of samples sent in one request
	public static final int MAX_BUNCH_SIZE = 60;
	
	// The thread running this instance
	protected Thread thread;
	
	// The preferences containing the data to upload
	protected SharedPreferences prefs;
	
	// Counts the number of unuploaded samples
	protected int dataCounter;
	
	// The URL to which to data is uploaded
	protected String uploadURL;
	
	// The authentication token that is given as a parameter in the upload requests
	protected String accessToken;
	
	/**
	 * Constructor
	 * 
	 * @param context Context used to get the preferences containing the data
	 */
	public DataUploader(ContextWrapper context, String uploadURL, String accessToken) {
		prefs = context.getSharedPreferences("org.fluxtream.flx_polar_h7_data", Context.MODE_PRIVATE);
		dataCounter = prefs.getAll().size();
		setParameters(uploadURL, accessToken);
	}
	
	public void setParameters(String uploadURL, String accessToken) {
		this.uploadURL = uploadURL;
		this.accessToken = accessToken;
	}
	
	/**
	 * Starts an independant thread running this instance. This should be called once.
	 */
	public synchronized void startThread() {
		thread = new Thread(this);
		thread.start();
	}
	
	/**
	 * Stops the thread running this instance
	 */
	public synchronized void stopThread() {
		if (thread != null) {
			thread.interrupt();
			thread = null;
		}
	}
	
	@Override
	public void run() {
		Log.i(PolarH7Service.LOG_TAG, "Starting HR upload thread");
		while (true) {
			try {
				// Record the number of data before waiting for 5 seconds to see if new data has arrived in the meantime
				int dataCountBeforeWait = dataCounter;
				// Wait for 5 seconds (unless data size is already worth uploading)
				if (dataCounter < BUNCH_SIZE) {
					synchronized (this) {
						// Wait for 5 seconds. This will get notified if data comes in.
						this.wait(5000);
					}
				}
				// Check if data should be uploaded (i.e. if no data has been added over the last 5 seconds, or the bunch size has been reached)
				if (dataCounter != 0 && (dataCounter == dataCountBeforeWait || dataCounter >= BUNCH_SIZE)) {
					// No data received for 5 seconds or data bunch size reached
					synchronizeNextDataBunch();
				}
			} catch (InterruptedException e) {
				// Thread interrupted, it must be stopped
				Log.i(PolarH7Service.LOG_TAG, "Stopping data uploader thread");
				return;
			} catch (Exception e) {
				Log.e(PolarH7Service.LOG_TAG, "Error on the sync thread: " + e.toString());
				ForgeApp.event("heartrate.uploadError");
				try {
					// Sleep for 20 seconds, to avoid looping over failing requests
					Log.i(PolarH7Service.LOG_TAG, "Pausing HR upload thread for 20 seconds");
					Thread.sleep(20000); 
				} catch (InterruptedException ie) {
					// Thread interrupted, it must be stopped
					Log.i(PolarH7Service.LOG_TAG, "Stopping data uploader thread");
					return;
				}
			}
		}
	}
	
	/**
	 * Adds the given data to the upload queue. This may trigger an upload if the bunch size is reached.
	 */
	public synchronized void addDataToUpload(int heartBeat, int beatSpacing) {
		Editor editor = prefs.edit();
		editor.putString("" + System.currentTimeMillis(), encodeData(heartBeat, beatSpacing));
		editor.apply();
		dataCounter++;
		Log.i(PolarH7Service.LOG_TAG, "Adding data (" + heartBeat + ", " + beatSpacing + ") ; Data counter = " + dataCounter);
		this.notify();
	}
	
	/**
	 * Encodes the data as a string to be stored
	 */
	protected String encodeData(int heartBeat, int beatSpacing) {
		return heartBeat + "," + beatSpacing;
	}
	
	/**
	 * Decodes the data from a string to a (heartBeat, beatSpacing) pair
	 */
	protected int[] decodeData(String data) {
		try {
			String[] strings = data.split(",");
			int[] ints = new int[2];
			if (ints.length >= 1) {
				ints[0] = Integer.parseInt(strings[0]);
			}
			if (ints.length >= 2) {
				ints[1] = Integer.parseInt(strings[1]);
			}
			return ints;
		} catch (Exception e) {
			Log.e(PolarH7Service.LOG_TAG, "Error while parsing data (" + data + ")");
			return new int[]{0, 0};
		}
	}
	
	/**
	 * Reads in the prefs the set of data to be uploaded next
	 */
	protected synchronized Map<String, int[]> getDataToSynchronize() {
		Map<String, int[]> data = new HashMap<String, int[]>();
		for (Entry<String, ?> entry : prefs.getAll().entrySet()) {
			String stringValue = (String)entry.getValue();
			data.put(entry.getKey(), decodeData(stringValue));
			if (data.size() >= MAX_BUNCH_SIZE) {
				break;
			}
		}
		return data;
	}
	
	/**
	 * Removes from the prefs the data when it has been uploaded
	 */
	protected synchronized void removeDataFromQueue(Map<String, ?> data) {
		Editor editor = prefs.edit();
		for (String key : data.keySet()) {
			editor.remove(key);
		}
		editor.apply();
		dataCounter = prefs.getAll().size();
	}
	
	/**
	 * Send an http request to upload the next bunch of data
	 * 
	 * @throws Exception
	 */
	protected void synchronizeNextDataBunch() throws Exception {
		Log.i(PolarH7Service.LOG_TAG, "Synchronizing now");
		ForgeApp.event("heartrate.startUpload");
		
		// Collect data to upload
		Map<String, int[]> data = getDataToSynchronize();
		
		// Construct data string
		JsonArray json = new JsonArray();
		for (String key : data.keySet()) {
			int[] ints = data.get(key);
			JsonArray dataArray = new JsonArray();
			dataArray.add(new JsonPrimitive(Long.parseLong(key) / 1000)); // Convert milliseconds to seconds
			dataArray.add(new JsonPrimitive(ints[0]));
			dataArray.add(new JsonPrimitive(ints[1]));
			json.add(dataArray);
			
		}
		String dataString = json.toString();
		Log.i(PolarH7Service.LOG_TAG, "Data to send: " + dataString);
		
		// Create list of parameters
		List<NameValuePair> params = new ArrayList<NameValuePair>();
		params.add(new BasicNameValuePair("dev_nickname", "PolarStrap"));
		params.add(new BasicNameValuePair("channel_names", "[\"HeartBeat\",\"BeatSpacing\"]"));
		params.add(new BasicNameValuePair("data", dataString));
		params.add(new BasicNameValuePair("access_token", accessToken));
		
		// Create request
		HttpPost httpPost = new HttpPost(uploadURL);
		httpPost.setEntity(new UrlEncodedFormEntity(params, HTTP.UTF_8));
		
		// Execute request with a timeout of 30 seconds
		HttpParams httpParams = new BasicHttpParams();
		HttpConnectionParams.setConnectionTimeout(httpParams, 30000);
		HttpConnectionParams.setSoTimeout(httpParams, 30000);
		HttpClient httpClient = new DefaultHttpClient(httpParams);
		HttpResponse response = httpClient.execute(httpPost);
		
		// Check response status code
		int statusCode = response.getStatusLine().getStatusCode();
		Log.i(PolarH7Service.LOG_TAG, "Response status code: " + statusCode);
		if (statusCode / 100 != 2) {
			throw new Exception("Wrong http response received when fetching access token (" + statusCode + ")");
		}
		// Read response
		BufferedReader reader = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));
		String line = null;
		String responseBody = "";
		while ((line = reader.readLine()) != null) {
			responseBody = responseBody + line;
		}
		Log.i(PolarH7Service.LOG_TAG, "Upload response: " + responseBody);
		// Parse response
		JSONObject jsonResponse = new JSONObject(responseBody);
		// Get whether the request was successful
		String successfulRecords = jsonResponse.getString("successful_records");
		boolean success = (Integer.parseInt(successfulRecords) != 0);
		if (success) {
			// The request was successful, remove data from queue
			removeDataFromQueue(data);
			if (dataCounter < BUNCH_SIZE) {
				ForgeApp.event("heartrate.uploadDone");
			}
		}
	}
	
}

File: modules/flx_polar_h7/inspector/ios-inspector/ForgeModule/DataUploader.m
//
//  DataUploader.m
//  ForgeModule
//
//  Created by Julien Dupuis on 14/01/15.
//  Copyright (c) 2015 Trigger Corp. All rights reserved.
//

#import "DataUploader.h"

@interface DataUploader()

// Upload thread
@property (strong, nonatomic) NSThread *thread;
@property (nonatomic, strong) NSCondition *interruptCondition;

// Upload request parameters
@property (strong, nonatomic) NSString *uploadURL;
@property (strong, nonatomic) NSString *accessToken;

// List of data that must still be synchronized
@property (strong, nonatomic) NSMutableDictionary *unsynchronizedData;

@end


@implementation DataUploader

- (instancetype)initWithUploadURL:(NSString *)uploadURL accessToken:(NSString *)accessToken {
    if (self = [super init]) {
        // Get user defaults containing the sampled data
        self.unsynchronizedData = [[[NSUserDefaults standardUserDefaults] valueForKey:@"heartrate.unsynchronizedData"] mutableCopy];
        if (!self.unsynchronizedData) self.unsynchronizedData = [NSMutableDictionary new];
        // Init parameters
        [self setParametersUploadURL:uploadURL accessToken:accessToken];
        // Create interrupt condition for the thread
        self.interruptCondition = [NSCondition new];
    }
    return self;
}

- (void)setParametersUploadURL:(NSString *)uploadURL accessToken:(NSString *)accessToken {
    self.uploadURL = uploadURL;
    self.accessToken = accessToken;
}

- (void)startThread {
    // Create and start the thread that will check for pictures
    self.thread = [[NSThread alloc] initWithTarget:self selector:@selector(runUploadThread) object:nil];
    [self.thread start];
}

- (void)stopThread {
    [self.thread cancel];
    // Interrupt waiting thread
    [self.interruptCondition lock];
    [self.interruptCondition signal];
    [self.interruptCondition unlock];
    // Set thread to null
    self.thread = NULL;
}

- (void)addDataToUploadHeartBeat:(int)heartBeat beatSpacing:(int)beatSpacing {
        // Add new data to
        NSUInteger dataSizeBefore = self.unsynchronizedData.count;
        [self.unsynchronizedData setValue:[DataUploader encodeDataHeartRate:heartBeat beatSpacing:beatSpacing]
                                   forKey:[NSNumber numberWithDouble:[[NSDate date] timeIntervalSince1970]].description];
        [[NSUserDefaults standardUserDefaults] setValue:self.unsynchronizedData forKey:@"heartrate.unsynchronizedData"];
        [[NSUserDefaults standardUserDefaults] synchronize];
        NSLog(@"Adding data (%d, %d) ; Data counter = %d", heartBeat, beatSpacing, (int)self.unsynchronizedData.count);
        // Interrupt waiting thread
        if (self.unsynchronizedData.count != dataSizeBefore) {
            [self.interruptCondition lock];
            [self.interruptCondition signal];
            [self.interruptCondition unlock];
        } else {
            // Data has been replaced by a new value, don't signal count change
        }
}

/**
 * Encodes data as a string
 */
+ (NSString *)encodeDataHeartRate:(int)heartBeat beatSpacing:(int)beatSpacing {
    return [NSString stringWithFormat:@"%d,%d", heartBeat, beatSpacing];
}

/**
 * Decodes data from a string. Returns an array containing two integers : heart rate and beat spacing.
 */
+ (NSArray *)decodeData:(NSString *)data {
    int heartBeat = 0;
    int beatSpacing = 0;
    @try {
        NSArray *strings = [data componentsSeparatedByString:@","];
        if (strings.count >= 1) {
            heartBeat = [[strings objectAtIndex:0] intValue];
        }
        if (strings.count >= 2) {
            beatSpacing = [[strings objectAtIndex:1] intValue];
        }
    } @catch (NSException *e) {
        NSLog(@"Error while parsing data (%@)", data);
        heartBeat = 0;
        beatSpacing = 0;
    }
    return @[[NSNumber numberWithInt:heartBeat], [NSNumber numberWithInt:beatSpacing]];
}

/**
 * Loops to upload the data when it comes in
 */
- (void)runUploadThread {
    NSLog(@"Starting HR upload thread");
    while (![NSThread currentThread].cancelled) {
        @try {
            // Record the number of data before waiting for 5 seconds to see if new data has arrived in the meantime
            NSUInteger dataCountBeforeWait = self.unsynchronizedData.count;
            // Wait for 5 seconds (unless data size is already worth uploading)
            if (self.unsynchronizedData.count < BUNCH_SIZE) {
                @synchronized(self) {
                    NSLog(@"Waiting for 5 seconds...");
                    // Wait for 5 seconds. This will get notified if data comes in.
                    [self.interruptCondition lock];
                    [self.interruptCondition waitUntilDate:[NSDate dateWithTimeIntervalSinceNow:5]];
                    [self.interruptCondition unlock];
                    NSLog(@"Waiting over");
                }
            }
            // Stop if the thread should be canceled
            if ([NSThread currentThread].cancelled) return;
            // Check if data should be uploaded (i.e. if no data has been added over the last 5 seconds, or the bunch size has been reached)
            if (self.unsynchronizedData.count != 0 && (self.unsynchronizedData.count == dataCountBeforeWait || self.unsynchronizedData.count >= BUNCH_SIZE)) {
                // No data received for 5 seconds or data bunch size reached
                NSLog(@"self.dataCounter = %d and dataCountBeforeWait = %d", (int)self.unsynchronizedData.count, (int)dataCountBeforeWait);
                [self synchronizeNextDataBunch];
            } else {
                NSLog(@"Not synchronizing yet");
            }
        } @catch (NSException *e) {
            // End the thread if it must be terminated
            if ([NSThread currentThread].cancelled) return;
            // An error has occurred, wait for 20 seconds before the next retry
            [[ForgeApp sharedApp] event:@"heartrate.uploadError" withParam:NULL];
            NSLog(@"Exception: %@", e);
            NSLog(@"Pausing HR upload thread for 20 seconds");
            NSCondition *condition = [NSCondition new];
            [condition lock];
            [condition waitUntilDate:[NSDate dateWithTimeIntervalSinceNow:20]];
            [condition unlock];
        }
    }
}

/**
 * Reads in the prefs the set of data to be uploaded next
 */
- (NSDictionary *)getDataToSynchronize {
    @synchronized(self) {
        NSMutableDictionary *data = [NSMutableDictionary new];
        for (NSString *key in self.unsynchronizedData) {
            [data setValue:self.unsynchronizedData[key] forKey:key];
            if (data.count >= MAX_BUNCH_SIZE) break;
        }
        return data;
    }
}

/**
 * Removes from the prefs the data when it has been uploaded
 */
- (void)removeDataFromQueue:(NSDictionary *)data {
    @synchronized(self) {
        for (NSString *key in data) {
            [self.unsynchronizedData removeObjectForKey:key];
        }
        [[NSUserDefaults standardUserDefaults] setValue:self.unsynchronizedData forKey:@"heartrate.unsynchronizedData"];
        [[NSUserDefaults standardUserDefaults] synchronize];
    }
}

/**
 * Uploads the next MAX_BUNCH_SIZE samples
 */
- (void)synchronizeNextDataBunch {
    NSLog(@"Synchronizing now");
    [[ForgeApp sharedApp] event:@"heartrate.startUpload" withParam:NULL];
    
    // Collect data to upload
    NSDictionary *data = self.getDataToSynchronize;
    
    // Construct data string
    NSMutableArray *dataArray = [NSMutableArray new];
    for (NSString *key in data) {
        NSArray *ints = [DataUploader decodeData:data[key]];
        [dataArray addObject:@[[NSNumber numberWithInt:key.intValue], ints[0], ints[1]]];
    }
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dataArray
                                                       options:0
                                                         error:&error];
    NSString *dataString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    NSLog(@"Data to send: %@", dataString);
    
    // Create request
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    [request setURL:[NSURL URLWithString:[self.uploadURL stringByAppendingString:@"?"]]];
    [request setTimeoutInterval:30.0];
    // Add body to request
    NSData *body = [[NSString stringWithFormat:@"access_token=%@&dev_nickname=%@&channel_names=%@&data=%@",
                     [self.accessToken stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding],
                     @"PolarStrap",
                     [@"[\"HeartBeat\",\"BeatSpacing\"]" stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding],
                     [dataString stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]]
                    dataUsingEncoding:NSUTF8StringEncoding];
    NSString *bodyLength = [NSString stringWithFormat:@"%ld", (long)[body length]];
    [request setValue:bodyLength forHTTPHeaderField:@"Content-Length"];
    [request setHTTPBody:body];
    
    // Send request and wait for response
    NSURLResponse *response;
    NSData *responseData = [NSURLConnection sendSynchronousRequest:request
                                                 returningResponse:&response
                                                             error:&error];
    
    // Make sure status code is 2**
    int statusCode = (int)[(NSHTTPURLResponse*) response statusCode];
    NSLog(@"Response status code: %d", statusCode);
    if (statusCode / 100 != 2) @throw [NSException exceptionWithName:@"Wrong status code"
                                                              reason:[NSString stringWithFormat:@"Received wrong status code: %d", statusCode]
                                                            userInfo:nil];
    
    // Make sure response indicates successful records
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:responseData options:NSJSONReadingAllowFragments error:&error];
    int successfulRecords = [NSString stringWithFormat:@"%@", [json objectForKey:@"successful_records"]].intValue;
    if (!successfulRecords) @throw [NSException exceptionWithName:@"Wrong response"
                                                           reason:[NSString stringWithFormat:@"Unexpected response: %@", [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding]]
                                                         userInfo:nil];
    
    // The request was successful, remove data from queue
    NSLog(@"Removing %d pieces of data", (int)data.count);
    [self removeDataFromQueue:data];
    if (self.unsynchronizedData.count < BUNCH_SIZE) {
        [[ForgeApp sharedApp] event:@"heartrate.uploadDone" withParam:NULL];
    }
}

@end
